[16.0][ADD] stock_valuation_specific_identification

matt454357 opened the pull request:
New module to add Specific Identification Inventory Valuation Method.

rousseldenis committed:
[12.0][ADD] stock_production_lot_quantity_tree

oca-travis committed:
[UPD] Update stock_production_lot_quantity_tree.pot

manuelcalerosolis committed:
[MIG] stock_production_lot_quantity_tree: Migration to 13.0

oca-travis committed:
[UPD] Update stock_production_lot_quantity_tree.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
[UPD] README.rst

AdrianaSaiz committed:
[IMP] stock_production_lot_quantity_tree: pre-commit stuff

AdrianaSaiz committed:
[MIG] stock_production_lot_quantity_tree: Migration to 16.0

matt454357 committed:
[16.0][ADD] stock_valuation_specific_identification

New module to add Specific Identification Inventory Valuation Method.

yostashiro commented:
I think this module shares a similar objective with https://github.com/OCA/stock-logistics-workflow/pull/1527.

One major difference, I understand, is that this module creates an SVL record per serial and the other one (stock_valuation_fifo_lot) does not (multiple lots/serials can be linked to an SVL record).  Do you see a conceptual issue with the approach of the other module?

matt454357 commented:
Hi @yostashiro,

Thanks for pointing out `stock_valuation_fifo_lot`. I had not noticed it.

I like the approach taken in `stock_valuation_fifo_lot`, but it doesn't correctly handle revaluations.

I believe `stock_valuation_specific_identification` better describes the module.  There are similar 3rd party modules with similar descriptions (https://apps.odoo.com/apps/modules/browse?search=specific+identification). However, if `stock_valuation_fifo_lot` is the preferred name, I can move the changes to that module.

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (185 of 185 strings)

Translation: stock-logistics-warehouse-16.0/stock-logistics-warehouse-16.0-stock_reserve
Translate-URL: https://translation.odoo-community.org/projects/stock-logistics-warehouse-16-0/stock-logistics-warehouse-16-0-stock_reserve/it/

Commit:
[UPD] Update stock_production_lot_quantity_tree.pot

OCA-git-bot committed:
[BOT] post-merge updates

AungKoKoLin1997 commented on stock_valuation_specific_identification/models/stock_lot.py:
```diff
@@ -0,0 +1,232 @@
+    _inherit = "stock.lot"
+
+    specific_ident_cost = fields.Boolean(
+        related="product_id.categ_id.property_specific_ident_cost",
```
```suggestion
        related="product_id.property_specific_ident_cost",
```

AungKoKoLin1997 commented on stock_valuation_specific_identification/models/product.py:
```diff
@@ -0,0 +1,154 @@
+class ProductProduct(models.Model):
+    _inherit = "product.product"
+
+    specific_ident_cost = fields.Boolean(
```
I failed to understand to add this field in `product.product` model with `readonly` instead of just using from `product.template`

AungKoKoLin1997 commented on stock_valuation_specific_identification/models/stock_lot.py:
```diff
@@ -0,0 +1,232 @@
+        """Prepare the values for a stock valuation layer created by a delivery.
+        This should only be called for specific identification valuation method.
+
+        :param quantity: the quantity to value, expressed in `product_id..uom_id`
```
```suggestion
        :param quantity: the quantity to value, expressed in `product_id.uom_id`
```

AungKoKoLin1997 commented on stock_valuation_specific_identification/models/stock_lot.py:
```diff
@@ -0,0 +1,232 @@
+        company_id = self.env.context.get("force_company", self.env.company.id)
+        company = self.env["res.company"].browse(company_id)
+        quantity = -1 * quantity  # negative for out valuation layers
+        sid_vals = self._run_out_spec_ident(abs(quantity), company)
```
```suggestion
        sid_vals = self._run_out_spec_ident(quantity, company)
```

AungKoKoLin1997 commented on stock_valuation_specific_identification/models/stock_lot.py:
```diff
@@ -0,0 +1,232 @@
+            "product_id": self.product_id.id,
+            "value": sid_vals.get("value"),
+            "unit_cost": sid_vals.get("unit_cost"),
+            "quantity": quantity,
```
```suggestion
            "quantity": -1 * quantity,
```

chienandalu committed:
[FIX] stock_vlm_mgmt: location general inventory

We were lacking the default location id so we could select the tray. It
was also needed to avoid compting a pos when no tray is selected.

OCA-git-bot committed:
[BOT] post-merge updates

AungKoKoLin1997 commented on stock_valuation_specific_identification/models/product.py:
```diff
@@ -0,0 +1,154 @@
+                                % candidate.stock_move_id.name
+                            )
+                        )
+                    for (
+                        candidate_line
+                    ) in candidate.stock_move_id.move_line_ids.filtered(
+                        lambda x: x.lot_id == lot and x.remaining_qty > 0.0
+                    ):
+                        qty_taken_on_candidate = min(
+                            qty_to_take_on_candidates, candidate_line.remaining_qty
+                        )
```
I don't understand this part. `candidate_line` is an instance of `stock move line`, isn't? I don't think `stock move line` record has `remaining_qty` field. Am I missing something?

AungKoKoLin1997 commented on stock_valuation_specific_identification/models/stock_lot.py:
```diff
@@ -0,0 +1,232 @@
+
+        # Find back incoming stock valuation layers (called candidates here) to
+        # value `quantity`.
+        qty_to_take_on_candidates = quantity
```
Is this for the purpose of code readable?

AungKoKoLin1997 reviewed (commented):
Partial code review.

AungKoKoLin1997 commented:
@matt454357  There are some changes and commits that are not related to the PR. It seems like something went wrong during your rebase.

AungKoKoLin1997 commented on stock_valuation_specific_identification/models/stock_move.py:
```diff
@@ -0,0 +1,181 @@
+            and m.product_id.tracking != "none"
+        )
+        normal_moves = self - specific_ident_moves
+        recs = self.env["stock.valuation.layer"]
+        recs |= super(StockMove, normal_moves)._create_out_svl(
+            forced_quantity=forced_quantity
+        )
```
Isn't it enough?
```suggestion
        recs = super(StockMove, normal_moves)._create_out_svl(
            forced_quantity=forced_quantity
        )
```

AungKoKoLin1997 commented:
@matt454357 I decided to go with `stock_valuation_fifo_lot` for two reasons:

- The community is more likely to accept an existing module than a new one.
- In my opinion, this module's approach is simpler and involves fewer complete overrides of Odoo's standard functions.

I will create a new PR with some adjustments since the author of the existing PR has not followed up on it. Would you be interested in contributing your ideas (e.g., the revaluation process) after I create the PR?

matt454357 commented:
Hi @AungKoKoLin1997,

It has been a long time since I last did a PR. I shouldn't have merged the upstream changes made after starting my PR.  I thought it might fix a test that was failing in `stock_valuation_layer_accounting_date`. It seems `stock_valuation_layer_accounting_date` has a poorly written test that sometimes fails, depending on the timezone and time of day.

Anyway, I am happy to contribute to the `stock_valuation_fifo_lot` module. I would like to add more than just the revaluation stuff. For example, drop the `stock_no_negative` requirement, and enforce no-negative for specific identification moves only.

Revaluation is accessed from the valuation tree view, after grouping by lot/serial.  Therefore, we may want to change `stock_valuation_fifo_lot` to a single lot/serial per SVL record, rather than multiple per SVL record.

AungKoKoLin1997 commented:
> Anyway, I am happy to contribute to the `stock_valuation_fifo_lot` module. I would like to add more than just the revaluation stuff. For example, drop the `stock_no_negative` requirement, and enforce no-negative for specific identification moves only.

@matt454357 Thank you for your interest. I have created draft state PR https://github.com/OCA/stock-logistics-workflow/pull/1676 and I have to add test cases and consider fixing issue that is proposed https://github.com/OCA/stock-logistics-workflow/pull/1350#issuecomment-1840407086. Let me ping you when it is ready.
